src/internal/operations/backup_test.go

```diff
@@ -1187,3 +1197,791 @@ func (suite *BackupOpIntegrationSuite) TestNewBackupOperation() {
+				// additional non recoverable errors to recoverable slice.
+				return errs.Failure() != nil
+			},
+		},
+		{
+			name: "fail after recovery, recoverable errors",
+			collFunc: func() []data.BackupCollection {
+				bc := []data.BackupCollection{
+					makeBackupCollection(
+						tmp,
+						locPath,
+						[]odMock.Data{
+							makeODMockData("file1", nil, time.Now(), false, nil),
+							makeODMockData("file2", nil, time.Now(), false, assert.AnError),
+						}),
+				}
+
+				return bc
+			},
+			metadataCollFunc: func() []graph.MetadataCollectionEntry {
+				return makeMetadataCollectionEntries("url/1", driveID, folderID, tmp)
+			},
+			failurePolicy:   control.FailAfterRecovery,
+			expectRunErr:    assert.Error,
+			expectBackupTag: model.AssistBackup,
+			expectFaults: func(errs *fault.Bus) bool {
+				return errs.Failure() != nil && len(errs.Recovered()) > 0
+			},
+		},
+	}
+	for _, test := range table {
+		suite.Run(test.name, func() {
+			t := suite.T()
+
+			ctx, flush := tester.NewContext(t)
+			defer flush()
+
+			cs := test.collFunc()
+
+			mc, err := graph.MakeMetadataCollection(
+				tenantID,
+				userID,
+				path.OneDriveService,
+				path.FilesCategory,
+				test.metadataCollFunc(),
+				func(*support.ControllerOperationStatus) {})
+			require.NoError(t, err, clues.ToCore(err))
+
+			cs = append(cs, mc)
+			bp := &mockBackupProducer{
+				colls:                   cs,
+				injectNonRecoverableErr: test.injectNonRecoverableErr,
+			}
+
+			bo, err := NewBackupOperation(
+				ctx,
+				opts,
+				suite.kw,
+				suite.sw,
+				bp,
+				acct,
+				osel.Selector,
+				selectors.Selector{DiscreteOwner: userID},
+				evmock.NewBus())
+			require.NoError(t, err, clues.ToCore(err))
+
+			bo.Options.FailureHandling = test.failurePolicy
+
+			err = bo.Run(ctx)
+			test.expectRunErr(t, err, clues.ToCore(err))
+			// TODO: check status
+			if len(test.expectBackupTag) > 0 {
+				bID := bo.Results.BackupID
+				require.NotEmpty(t, bID)
+
+				bup := backup.Backup{}
+				err := suite.ms.Get(ctx, model.BackupSchema, bID, &bup)
+				require.NoError(t, err, clues.ToCore(err))
+
+				require.Equal(t, test.expectBackupTag, bup.Tags[model.BackupTypeTag])
+			}
+
+			require.True(t, test.expectFaults(bo.Errors))
+		})
+	}
+}
+
+func selectFilesFromDeets(d details.Details) map[string]details.Entry {
+	files := make(map[string]details.Entry)
+
+	for _, ent := range d.Entries {
+		if ent.Folder != nil {
+			continue
+		}
+
+		files[ent.ItemRef] = ent
+	}
+
+	return files
+}
+
+// TestExtensionsIncrementals tests presence of corso extension data in details
+// Note that since we are mocking out backup producer here, corso extensions can't be
+// attached as they would in prod. However, this is fine here, since we are more interested
+// in testing whether deets get carried over correctly for various scenarios.
+func (suite *AssistBackupIntegrationSuite) TestExtensionsIncrementals() {
```

Comment:
I love seeing these tests. It's great to have something that can check some of the more complex behavior of the system!